src/grid_interpolator.cc

```diff
@@ -47,22 +59,35 @@ GridInterpolator::~GridInterpolator()
 
 double GridInterpolator::operator()(double x, double y, double z) const
     {
-    const double fx = x/dx_;
-    const double fy = y/dy_;
-    const double fz = z/dz_;
-    const int bin_x = std::floor(fx); 
-    const int bin_y = std::floor(fy); 
-    const int bin_z = std::floor(fz); 
+    const double fx = (x - std::get<0>(lower_bounds_))/dx_;
+    const double fy = (y - std::get<1>(lower_bounds_))/dy_;
+    const double fz = (z - std::get<2>(lower_bounds_))/dz_;
+    
+    int bin_x = std::floor(fx);
+    int bin_y = std::floor(fy);
+    int bin_z = std::floor(fz);
+    
+    const auto bounds = n_.getBounds();
+    if(fx == std::get<0>(bounds) - 1 && x == std::get<0>(upper_bounds_))
+       {
+       --bin_x;
+       }
+    if(fy == std::get<1>(bounds) - 1 && x == std::get<1>(upper_bounds_))
+       {
+       --bin_y;
+       }
+    if(fz == std::get<2>(bounds) - 1 && x == std::get<2>(upper_bounds_))
+       {
+       --bin_z;
+       }
```

Comment:
For readability, I would interleave these like this and also add a comment on the first one about why we decrement:
```suggestion

    const auto bounds = n_.getBounds();
    // x
    int bin_x = std::floor(fx);
    if(fx == std::get<0>(bounds) - 1 && x == std::get<0>(upper_bounds_))
       {
       /* for points exactly on the upper boundary, decrease by 1 to put it in
          bin below. this makes sure we only access bins within the bounds
          during interpolation
       */
       --bin_x;
       }
    // y
    int bin_y = std::floor(fy);
    if(fy == std::get<1>(bounds) - 1 && x == std::get<1>(upper_bounds_))
       {
       --bin_y;
       }
    // z
    int bin_z = std::floor(fz);
    if(fz == std::get<2>(bounds) - 1 && x == std::get<2>(upper_bounds_))
       {
       --bin_z;
       }
```